### PR TITLE
LC submission time check UTC fix

### PR DESCRIPTION
**Pull request by kk-min**
Fixes to LC submission time check.

# Changes

- Fix LC submission time check by considering UTC values (midnight to midnight), translates to 8am to 8am in SGT
- Refactor LC submission time check into a function (/utils/isValidLCDate.js) for easier testing
- Add Jest testing framework to project + unit tests for LC date validation function
- Add Github actions script (by @MrMarciaOng) to automatically run Jest tests on push to branches/PR to main

### CODE REVIEW

File: utils/isValidLCDate.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+const dayjs = require('dayjs');
+let isBetween = require('dayjs/plugin/isBetween');
+let utc = require('dayjs/plugin/utc');
+dayjs.extend(isBetween);
+dayjs.extend(utc);
+
+function isValidLCDate(...args) {
```
Comment (paradite) [inline utils/isValidLCDate.js]: This function is very confusing.

Can you name the parameters explicitly and clearly? Make a distinction between date being submitted and actual date of submission.

File: test/isValidLCDate.test.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
+const isValidLCDate = require('../utils/isValidLCDate');
+
+let mockDate;
+let mockHour;
+
+test('Valid LC Date', () => {
+  mockDate = '20210121';
```
Comment (paradite) [inline test/isValidLCDate.test.js]: name `mockDate` and `mockHour` explicitly following the parameter names in `isValidLCDate`.

File: utils/isValidLCDate.js (hunk 1 above)
Comment (paradite) [reply utils/isValidLCDate.js]: You can have more parameters in a function than parameters actually being passed. they will just become undefined.

**Review by paradite — CHANGES_REQUESTED** [review]
left some comments

File: test/isValidLCDate.test.js (hunk 2 above)
Comment (kk-min) [reply test/isValidLCDate.test.js:7]: Just to clarify you mean that the variable name here in the test should correspond to the explicit parameter name in isValidLCDate?

**Comment by kk-min** [comment]
Changed function definition as requested.

**Comment by paradite** [comment]
Please name the parameters explicitly and clearly on what they mean. 

Please read my comments again.

**Comment by kk-min** [comment]
Ok I see where the confusion might be coming from. Just a quick breakdown of the function with updated names:

- isValidLCDate() takes in a single string (**submittedDateString**) that corresponds to the YYYYMMDD the user submitted, and returns a boolean value that determines whether it is within the valid LC interval **based on the current server time**.
- This calculation is done by creating **leftBound** and **rightBound** (UTC midnight of the current server date and the next server day respectively) by calling dayjs.utc() to get the current "now" of the server and adjusting the time to midnight, and then checking if the submitted date with the current server hour falls in-between the two bounds.
- However, for testing purposes we wish to mock this server date. If we pass in 2 more arguments, **mockServerDate** and **mockServerHour**, these two arguments will be used to create the **leftBound** and **rightBound** instead of using the actual server datetime, effectively mocking the server date for this function.
- I have tried all manners of ways via Jest to mock this date rather than mocking it within the function itself (fake timers, jest.mock, spys, etc.), but it is extremely unreliable; the only way that it works using Jest's built in features is a [manual mock](https://jestjs.io/docs/manual-mocks), but that involves putting the dayjs.utc() function in a separate directory and we can no longer easily adjust the mock server date in between tests.

**Comment by paradite** [comment]
"However, for testing purposes we wish to mock this server date. If we pass in 2 more arguments, mockServerDate and mockServerHour, these two arguments will be used to create the leftBound and rightBound instead of using the actual server datetime, effectively mocking the server date for this function."

Check your code again, your hour is not being used for left bound and right bound.

**Comment by paradite** [comment]
Please keep variable names consistently throughout the code. Don't change names in the middle or reassign them. It causes confusions and bugs.

**Comment by kk-min** [comment]
That is correct, pardon the inaccuracy in my comment. We don't use the hour for leftBound and rightBound because they are supposed to be at midnight UTC. mockServerHour is used for submissionDate (i.e. assigned to submissionHour).

**Comment by paradite** [comment]
> That is correct, pardon the inaccuracy in my comment. We don't use the hour for leftBound and rightBound because they are supposed to be at midnight UTC. mockServerHour is used for submissionDate (i.e. assigned to submissionHour).

Yes. Please rename your variable names and parameters accordingly to reflect their actual meanings and keep them consistent throughout.

**Comment by paradite** [comment]
Your "mockServerHour" is not server hour, but user submission hour. Calling it "mockServerHour" is not accurate and causes confusion and bugs.

**Comment by kk-min** [comment]
The reason I used "mockServerHour" is because the user submission hour is actually **derived from the server's hour**. When users submit, they only declare YYYYMMDD, they don't declare the hour. So the hour of submission is from the server's time itself, which when being mocked is of course the "mock" server hour.

So during the tests, we pass in "mockServerHour" to make it clear that we are simulating the server's hour. Inside the function, we assign to submissionHour to signify that we are taking the server's hour (actual or mock) as the hour of user submission.

^ Can you consider this and if it's not clear or acceptable I'll change it to something like "userSubmissionHour", but I think it might make the tests file confusing...

**Comment by paradite** [comment]
I suggest your rewrite the logic to remove the confusing naming. 

Make clear distinctions between user submitted date time and actual date time when the submission happens, and test that they are working as expected.

**Comment by kk-min** [comment]
After thinking about this a bit further I decided to remove mockServerHour completely. Turns out we don't really need it after all; it was necessary if we considered the server time in SGT (as we have to make sure it is within the 8am interval), but since we are now using UTC, we don't need to know the hour of submission; as long as the user input YYYYMMDD is the same as the current (or mocked) server date, we can conclude that it is a valid submission.

**Comment by paradite** [comment]
I don't think this is actually needed or now? Don't want to introduce many changes.

**Comment by kk-min** [comment]
Sure, that is up to your discretion. However I think right now the interval check is not working as expected since we manually set the time to 8 in SGT.

Suppose user submits #LC20230210 on date 20230210 at time 3 am. By right the entry should be #LC20230209 since it's not 8am SGT yet. But it will still go through as successful. See line **279** [here](https://github.com/paradite/16x-bot/commit/9fc3043a7466ebd947e6e16b09b7c3ccc878b7f5) for the reason.

**Pull request closed without merging**